axum-macros/src/from_request/attr.rs

```diff
@@ -0,0 +1,202 @@
+
+    let mut out = FromRequestContainerAttr::default();
+
+    for from_request_attr in attrs {
+        match from_request_attr {
+            ContainerAttr::Via { via, path } => {
+                if out.via.is_some() {
+                    return Err(double_attr_error("via", via));
+                } else {
+                    out.via = Some((via, path));
+                }
+            }
+            ContainerAttr::RejectionDerive {
+                rejection_derive,
+                opt_outs,
+            } => {
+                if out.rejection_derive.is_some() {
+                    return Err(double_attr_error("rejection_derive", rejection_derive));
+                } else {
+                    out.rejection_derive = Some((rejection_derive, opt_outs));
+                }
+            }
+        }
+    }
+
+    Ok(out)
+}
+
+pub(crate) fn parse_attrs<T>(attrs: &[syn::Attribute]) -> syn::Result<Punctuated<T, Token![,]>>
+where
+    T: Parse,
+{
+    let attrs = attrs
+        .iter()
+        .filter(|attr| attr.path.is_ident("from_request"))
+        .map(|attr| attr.parse_args_with(Punctuated::<T, Token![,]>::parse_terminated))
+        .collect::<syn::Result<Vec<_>>>()?
+        .into_iter()
+        .flatten()
+        .collect::<Punctuated<T, Token![,]>>();
+    Ok(attrs)
+}
+
+fn double_attr_error<T>(ident: &str, spanned: T) -> syn::Error
+where
+    T: ToTokens,
+{
+    syn::Error::new_spanned(spanned, format!("`{}` specified more than once", ident))
+}
+
+enum ContainerAttr {
+    Via {
+        via: kw::via,
+        path: syn::Path,
+    },
+    RejectionDerive {
+        rejection_derive: kw::rejection_derive,
+        opt_outs: RejectionDeriveOptOuts,
+    },
+}
+
+impl Parse for ContainerAttr {
+    fn parse(input: ParseStream) -> syn::Result<Self> {
+        let lh = input.lookahead1();
+        if lh.peek(kw::via) {
+            let via = input.parse::<kw::via>()?;
+            let content;
+            syn::parenthesized!(content in input);
+            content.parse().map(|path| Self::Via { via, path })
+        } else if lh.peek(kw::rejection_derive) {
+            let rejection_derive = input.parse::<kw::rejection_derive>()?;
+            let content;
+            syn::parenthesized!(content in input);
+            content.parse().map(|opt_outs| Self::RejectionDerive {
+                rejection_derive,
+                opt_outs,
+            })
+        } else {
+            Err(lh.error())
+        }
+    }
+}
+
+enum FieldAttr {
+    Via { via: kw::via, path: syn::Path },
+}
+
+impl Parse for FieldAttr {
+    fn parse(input: ParseStream) -> syn::Result<Self> {
+        let lh = input.lookahead1();
+        if lh.peek(kw::via) {
+            let via = input.parse::<kw::via>()?;
+            let content;
+            syn::parenthesized!(content in input);
+            content.parse().map(|path| Self::Via { via, path })
+        } else {
+            Err(lh.error())
+        }
+    }
+}
```

Comment:
The duplication in this parsing code bothers me a little but I couldn't find a nicer solution that gave good error messages.

Reply:
Seems fine to me as-is.